test/test_testing.py

```diff
@@ -2396,6 +2396,77 @@ def test_begin_with_initial_hooks_install_sets_status(self):
+        self.addCleanup(harness.cleanup)
+
+        harness.set_leader(True)
+        harness.begin_with_initial_hooks()
+        harness.reinitialize_charm()
+        self.assertTrue(harness.charm.unit.is_leader())
+
+        harness.set_leader(False)
+        harness.reinitialize_charm()
+        self.assertFalse(harness.charm.unit.is_leader())
+
+    def test_reinitialize_retains_peer_relations(self):
+        class PeerCharm(RelationEventCharm):
+            def __init__(self, framework):
+                super().__init__(framework)
+                self.observe_relation_events('peer')
+        harness = Harness(PeerCharm, meta='''
+            name: test-app
+            peers:
+              peer:
+                interface: app-peer
+            ''')
+        self.addCleanup(harness.cleanup)
+        harness.set_leader()
+        harness.add_relation('peer', 'test-app')
+        harness.begin_with_initial_hooks()
+        relations_before = harness._backend._relation_names.copy()
+        harness.reinitialize_charm()
+        self.assertEqual(relations_before, harness._backend._relation_names)
+
+    def test_reinitialize_retains_regular_relations(self):
+        class CharmWithDB(RelationEventCharm):
+            def __init__(self, framework):
+                super().__init__(framework)
+                self.observe_relation_events('db')
+        harness = Harness(CharmWithDB, meta='''
+            name: test-app
+            requires:
+              db:
+                interface: sql
+            ''')
+        self.addCleanup(harness.cleanup)
+        harness.set_leader()
+        rel_id = harness.add_relation('db', 'postgresql')
+        harness.add_relation_unit(rel_id, 'postgresql/0')
+        harness.update_relation_data(rel_id, 'postgresql/0', {'new': 'data'})
+        harness.update_relation_data(rel_id, 'postgresql', {'app': 'data'})
+        harness.begin_with_initial_hooks()
+        harness.reinitialize_charm()
+
+        # Pre-existing relations must be retained
+        self.assertEqual(harness._backend._relation_names, {rel_id: 'db'})
+
+        # Pre-existing relation data must be retained
+        self.assertEqual(harness.get_relation_data(rel_id, 'postgresql/0'), {'new': 'data'})
+        self.assertEqual(harness.get_relation_data(rel_id, 'postgresql'), {'app': 'data'})
+
+    def test_reinitialize_retains_status(self):
+        harness = Harness(RecordingCharm, meta='''name: test-app''')
+        self.addCleanup(harness.cleanup)
+        harness.set_leader()
+        harness.begin_with_initial_hooks()
+
+        harness.charm.unit.status = ActiveStatus("unit before reinit")
+        harness.charm.app.status = ActiveStatus("app before reinit")
+        harness.reinitialize_charm()
+        self.assertEqual(harness.charm.unit.status, ActiveStatus("unit before reinit"))
+        self.assertEqual(harness.charm.app.status, ActiveStatus("app before reinit"))
+
```

Comment:
Might be good to add a test with some __init__ logic affecting the charm state, like I pseudocoded in the first comment.